// @opencampus/ocid-connect-js/lib/sdk/lib/TransactionManager.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
/*!
* Copyright 2024-Present Animoca Brands Corporation Ltd. 
* 
* Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
* 
* The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
* 
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

// TransactionManager manages data that's stored over a "transaction"
// of the login cycle
// A transaction contains a few things that we want to carry over multiple
// times interacting with the SDK
// Currently just used for
// codeChallenge, codeVerifier, codeChallengeMethod

var TRANSACTION_STORAGE_NAME = 'oc-transaction-storage';
var TransactionManager = /*#__PURE__*/function () {
  function TransactionManager(StorageManagerClass) {
    _classCallCheck(this, TransactionManager);
    _defineProperty(this, "storageManager", void 0);
    this.storageManager = new StorageManagerClass(TRANSACTION_STORAGE_NAME);
  }

  // called when we start a new transaction
  return _createClass(TransactionManager, [{
    key: "clear",
    value: function clear() {
      this.storageManager.getStorageObject().clearStorage();
    }
  }, {
    key: "save",
    value: function save(meta) {
      this.storageManager.getStorageObject().setStorage(meta);
    }
  }, {
    key: "getTransactionMeta",
    value: function getTransactionMeta() {
      return this.storageManager.getStorageObject().getStorage();
    }
  }, {
    key: "hasActiveTransaction",
    value: function hasActiveTransaction() {
      var meta = this.storageManager.getStorageObject();
      return !!meta.codeChallenge && !!meta.codeVerifier;
    }
  }]);
}();
var _default = exports["default"] = TransactionManager;
